### ISSUE TITLE
Adjust more code for Python 3.5+

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#3020]
Since we dropped Python 3.4, there's some more 3.5 stuff we might be interested in:

- [ ] [`os.scandir()`](https://docs.python.org/3/library/os.html#os.scandir)
- [ ] [`subprocess.run()`](https://docs.python.org/3/library/subprocess.html#subprocess.run)
- [ ] [Changes in `traceback` module](https://docs.python.org/3/whatsnew/3.5.html#traceback)
- [ ] [`start` parameter for `enum.Enum`](https://docs.python.org/3/library/enum.html#enum.Enum)
- [ ] [`HTTPStatus` enum](https://docs.python.org/3/library/http.html#http.HTTPStatus)
- [ ] [Changes in `inspect` module](https://docs.python.org/3/whatsnew/3.5.html#inspect)
- [ ] [`exc_info` support in `logging`](https://docs.python.org/3/whatsnew/3.5.html#logging)
- [ ] [`signal.set_wakeup_fd()` on Windows](https://docs.python.org/3/library/signal.html#signal.set_wakeup_fd)
- [ ] [Unpacking generalizations](https://docs.python.org/3/whatsnew/3.5.html#whatsnew-pep-448)

**Comment by gyorb** [comment]
Hi, I would like to work on these improvements.
Is there any further useful info I should check before I get started?

**Comment by NoctuaNivalis** [comment]
Everything should be in the [contribution guidelines](https://github.com/qutebrowser/qutebrowser/blob/master/doc/contributing.asciidoc). Thank you for interest!

**Comment by The-Compiler** [comment]
Also, please note that it's preferable to have multiple smaller pull requests rather than one big one for those things. 😄

**Comment by gyorb** [comment]
Thanks, I will check out the contribution guidelines.
I planed to create separate pull requests for each of the items in the list.

**Comment by gebulmer** [comment]
As far as I can see, there's nowhere we would benefit from scandir over listdir. Every case I've seen doesn't use any other attributes than the filename.

**Comment by gyorb** [comment]
I did not check yet where the `os.scandir()` could be used. Should it be removed from the list?

**Comment by gebulmer** [comment]
I personally think so, can't see how it'd be a benefit for any of the cases (but I could be wrong of course!)

**Comment by The-Compiler** [comment]
Right! I looked through them, and indeed I don't think it makes sense anywhere.

**Comment by jberglinds** [comment]
Hi, we're a small group of students from KTH (Sweden) that would like to help out with this as part of a school project. We're thinking of doing the HTTPStatus enum refactor. I assume it's still something that you want done? @The-Compiler 
Should we create a separate issue for that to claim, or do you just want us to discuss it here?

**Comment by The-Compiler** [comment]
Hey @jberglinds, cool to hear about that! 👍 The only thing you should need to change there is `tests/end2end/fixtures/webserver.py` (currently using the `http.client` constants) and `tests/end2end/fixtures/webserver_sub.py` (currently using hard-coded codes).

It's fine to use this issue for it (as many of the tasks here should be quite small), but if you need help, I'd encourage you to join the [IRC channel](https://github.com/qutebrowser/qutebrowser#getting-help) as it's easier to discuss things there.

**Comment by The-Compiler** [comment]
@jberglinds You asked me some questions in the IRC channel and then disconnected after my answers, so I'm not sure you got them - here they are again:

- \<jberglinds> Regarding the HTTPStatus issue, and the tests. Is there a reason for only including tests for two separate paths?
- \<jberglinds> https://github.com/qutebrowser/qutebrowser/blob/master/tests/end2end/fixtures/test_webserver.py#L55-L59
- \<jberglinds> I would have expected all endpoints to have tests for them in a similar fashion.
- \<jberglinds> Also, about this code: https://github.com/qutebrowser/qutebrowser/blob/master/tests/end2end/fixtures/webserver.py#L63-L98. It seems to me as if the endpoints in the path_to_statuses dict are only checked for Request objects of that actual path, so without tests, most of these won't be used. Correct or incorrect?
- \<The-Compiler> jberglinds: so what you're looking at here is code inside qutebrowser's testsuite - the tests in tests/end2end spin up that webserver and make requests against it to test certain behavior
- \<The-Compiler> jberglinds: test_webserver.py is kind of a meta thing ;) It contains tests for that test-webserver. It's somewhat lightly tested because if there was something wrong with it, the "real" tests (which test qutebrowser) would fail anyways
- \<The-Compiler> jberglinds: so those endpoints are actually used, by the qutebrowser tests in tests/end2end

**Comment by jberglinds** [comment]
Thank you for posting here @The-Compiler . I had to leave before I got it.

----
Okay, I see. What I did was changing some status codes in the `webserver_sub.py` and running the `test_webserver.py` test (which didn't fail from that, so that's why I thought it wasn't tested). 
Now when I grep'ed for where a changed endpoint was used and ran that test, it failed like it should have.

Do you mind having some endpoint tests in `test_webserver.py` or do you want to keep it light like this? I added some quick tests there for the endpoints that would be changed for the enum. I get that this HTTPStatus rewrite is small enough to not cause any issues, but for the sake of our course the tests help.

**Comment by The-Compiler** [comment]
Go ahead! I don't think it's a good idea to add tests for *all* endpoints there (as that means there's now three places to change if some status changes for some reason) - but if you write a couple of tests which were helpful to you and adds an easy way to test this, it definitely doesn't hurt to add them! 👍

**Comment by The-Compiler** [comment]
New possible improvements for Python 3.6+

- [x] `WebKitElements` and `WebEngineElements` probably don't need to override `__init__` anymore just so that mypy knows the type of `self._tab`.
- [x] Same for `WebKitCaret` and `WebEngineCaret`
- [x] Same for `WebKitElement` and `WebEngineElement`
- [ ] Use [pyupgrade](https://github.com/asottile/pyupgrade) for f-strings (might be a good idea to hold off a bit due to possible conflicts with existing PRs)
- [ ] Use [com2ann](https://github.com/ilevkivskyi/com2ann) for type annotations (@stimut started working on this in #5829)
- [x] Drop conditionals around `secrets` module
- [ ] Various simplifications and proper annotations (rather than strings) around the `typing` module
- [ ] Various `str(...)` around `pathlib.Path` and `py.path.local` objects which can be dropped
- [x] `enum.auto()` for enums

**Comment by The-Compiler** [comment]
No reason to do those changes for 2.0.0 - while they should happen at some point, they're not very important.

**Comment by The-Compiler** [comment]
Switching to fstrings using `pyupgrade` is something we should do for 3.0.0 with all the other code churn (#1455).
